[inductor][5/N] triton support post-#5512, fix 1 and None handling

davidberard98 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #145515

This fixes handling for "1" and "None" args with new Triton versions. TL;DR: triton_meta["constants"] (which is passed to ASTSource) should be a map of {"kwarg_name": constant_value} for values which are tl.constexpr, or have a value of 1 or None (i.e. "specialized" constants). For constant args, triton_meta["signature"][arg_name] should be "constexpr" (even for specialized constants).

Note: This adds support for Triton versions after 5512; but not for versions in between 5220 and 5512 (i.e. `TritonAttrsDescriptorVersion.V3_BACKENDS_TUPLE`). There's a completely different format for constants/signature in the commit range in between.

To test: I ran `test_torchinductor.py` and `test_triton_kernels.py` with the main branch of triton (~jan 27). The only failing tests are aoti-related tests (which need to be fixed as a follow-up), and test_mutable_custom_op_fixed_layout2_cuda (which is failing with or without the new triton version on my machine); additionally, the split-scan/split-reduction kernels rely on https://github.com/triton-lang/triton/pull/5723.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

davidberard98 committed:
[WIP][inductor][3/N] triton support post-#5512, fix 1 and None handling

In the previous PRs, I removed equal_to_1 args from the "constants" (i.e. the "constants" arg passed to the ASTSource). However, we actually should add equal_to_1 args to "constants" (after PR 5220, equal_to_1 args were removed; but then after PR 5512, equal_to_1 and None args were added back to the "constants" arg).

[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP][inductor][3/N] triton support post-#5512, fix 1 and None handling"


This fixes handling for "1" and "None" args. TL;DR: triton_meta["constants"] (which is passed to ASTSource) should be a map of {"kwarg_name": "constant value"} for values which are tl.constexpr, or have a value of 1 or None (i.e. "specialized" constants). For constant args, triton_meta["signature"][arg_name] should be "constexpr" (even for specialized constants).

In a bit more detail:
* Triton commit 5220 actually removes "1" and "None" args from the "constants" argument to ASTSource, but adds tl.constexpr args into the signature
* Triton commit 5512 brings "1" and "None" args back into the "constants" argument, and leaves tl.constexpr args in the signature. But now, the "1" and "None" args in the signature are marked as "constexpr".

So if we want to implement support for the range of triton commits labeled `TritonAttrsDescriptorVersion.V3_BACKENDS_TUPLE`, this is something we should keep in mind.

...

To test this, I added this change to triton_heuristics.py:

```diff
diff --git a/torch/_inductor/runtime/triton_heuristics.py b/torch/_inductor/runtime/triton_heuristics.py
index 0b0ba7cff16..69330571b61 100644
--- a/torch/_inductor/runtime/triton_heuristics.py
+++ b/torch/_inductor/runtime/triton_heuristics.py
@@ -470,6 +470,12 @@ class CachingAutotuner(KernelInterface):
         if not ASTSource:
             raise RuntimeError("Installed triton version too old, please upgrade")

+        for k in compile_meta["constants"].keys():
+            assert (
+                k in compile_meta["signature"]
+                and compile_meta["signature"][k] == "constexpr"
+            )
+
         compile_args = (
             ASTSource(
                 self.fn,
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP][inductor][5/N] triton support post-#5512, fix 1 and None handling"


This fixes handling for "1" and "None" args. TL;DR: triton_meta["constants"] (which is passed to ASTSource) should be a map of {"kwarg_name": "constant value"} for values which are tl.constexpr, or have a value of 1 or None (i.e. "specialized" constants). For constant args, triton_meta["signature"][arg_name] should be "constexpr" (even for specialized constants).

In a bit more detail:
* Triton commit 5220 actually removes "1" and "None" args from the "constants" argument to ASTSource, but adds tl.constexpr args into the signature
* Triton commit 5512 brings "1" and "None" args back into the "constants" argument, and leaves tl.constexpr args in the signature. But now, the "1" and "None" args in the signature are marked as "constexpr".

So if we want to implement support for the range of triton commits labeled `TritonAttrsDescriptorVersion.V3_BACKENDS_TUPLE`, this is something we should keep in mind.

...

To test this, I added this change to triton_heuristics.py:

```diff
diff --git a/torch/_inductor/runtime/triton_heuristics.py b/torch/_inductor/runtime/triton_heuristics.py
index 0b0ba7cff16..69330571b61 100644
--- a/torch/_inductor/runtime/triton_heuristics.py
+++ b/torch/_inductor/runtime/triton_heuristics.py
@@ -470,6 +470,12 @@ class CachingAutotuner(KernelInterface):
         if not ASTSource:
             raise RuntimeError("Installed triton version too old, please upgrade")

+        for k in compile_meta["constants"].keys():
+            assert (
+                k in compile_meta["signature"]
+                and compile_meta["signature"][k] == "constexpr"
+            )
+
         compile_args = (
             ASTSource(
                 self.fn,
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP][inductor][5/N] triton support post-#5512, fix 1 and None handling"


This fixes handling for "1" and "None" args. TL;DR: triton_meta["constants"] (which is passed to ASTSource) should be a map of {"kwarg_name": "constant value"} for values which are tl.constexpr, or have a value of 1 or None (i.e. "specialized" constants). For constant args, triton_meta["signature"][arg_name] should be "constexpr" (even for specialized constants).

In a bit more detail:
* Triton commit 5220 actually removes "1" and "None" args from the "constants" argument to ASTSource, but adds tl.constexpr args into the signature
* Triton commit 5512 brings "1" and "None" args back into the "constants" argument, and leaves tl.constexpr args in the signature. But now, the "1" and "None" args in the signature are marked as "constexpr".

So if we want to implement support for the range of triton commits labeled `TritonAttrsDescriptorVersion.V3_BACKENDS_TUPLE`, this is something we should keep in mind.

...

To test this, I added this change to triton_heuristics.py:

```diff
diff --git a/torch/_inductor/runtime/triton_heuristics.py b/torch/_inductor/runtime/triton_heuristics.py
index 0b0ba7cff16..69330571b61 100644
--- a/torch/_inductor/runtime/triton_heuristics.py
+++ b/torch/_inductor/runtime/triton_heuristics.py
@@ -470,6 +470,12 @@ class CachingAutotuner(KernelInterface):
         if not ASTSource:
             raise RuntimeError("Installed triton version too old, please upgrade")

+        for k in compile_meta["constants"].keys():
+            assert (
+                k in compile_meta["signature"]
+                and compile_meta["signature"][k] == "constexpr"
+            )
+
         compile_args = (
             ASTSource(
                 self.fn,
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP][inductor][5/N] triton support post-#5512, fix 1 and None handling"


This fixes handling for "1" and "None" args. TL;DR: triton_meta["constants"] (which is passed to ASTSource) should be a map of {"kwarg_name": "constant value"} for values which are tl.constexpr, or have a value of 1 or None (i.e. "specialized" constants). For constant args, triton_meta["signature"][arg_name] should be "constexpr" (even for specialized constants).

In a bit more detail:
* Triton commit 5220 actually removes "1" and "None" args from the "constants" argument to ASTSource, but adds tl.constexpr args into the signature
* Triton commit 5512 brings "1" and "None" args back into the "constants" argument, and leaves tl.constexpr args in the signature. But now, the "1" and "None" args in the signature are marked as "constexpr".

So if we want to implement support for the range of triton commits labeled `TritonAttrsDescriptorVersion.V3_BACKENDS_TUPLE`, this is something we should keep in mind.

...

To test this, I added this change to triton_heuristics.py:

```diff
diff --git a/torch/_inductor/runtime/triton_heuristics.py b/torch/_inductor/runtime/triton_heuristics.py
index 0b0ba7cff16..69330571b61 100644
--- a/torch/_inductor/runtime/triton_heuristics.py
+++ b/torch/_inductor/runtime/triton_heuristics.py
@@ -470,6 +470,12 @@ class CachingAutotuner(KernelInterface):
         if not ASTSource:
             raise RuntimeError("Installed triton version too old, please upgrade")

+        for k in compile_meta["constants"].keys():
+            assert (
+                k in compile_meta["signature"]
+                and compile_meta["signature"][k] == "constexpr"
+            )
+
         compile_args = (
             ASTSource(
                 self.fn,
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP][inductor][5/N] triton support post-#5512, fix 1 and None handling"


This fixes handling for "1" and "None" args. TL;DR: triton_meta["constants"] (which is passed to ASTSource) should be a map of {"kwarg_name": "constant value"} for values which are tl.constexpr, or have a value of 1 or None (i.e. "specialized" constants). For constant args, triton_meta["signature"][arg_name] should be "constexpr" (even for specialized constants).

In a bit more detail:
* Triton commit 5220 actually removes "1" and "None" args from the "constants" argument to ASTSource, but adds tl.constexpr args into the signature
* Triton commit 5512 brings "1" and "None" args back into the "constants" argument, and leaves tl.constexpr args in the signature. But now, the "1" and "None" args in the signature are marked as "constexpr".

So if we want to implement support for the range of triton commits labeled `TritonAttrsDescriptorVersion.V3_BACKENDS_TUPLE`, this is something we should keep in mind.

...

To test this, I added this change to triton_heuristics.py:

```diff
diff --git a/torch/_inductor/runtime/triton_heuristics.py b/torch/_inductor/runtime/triton_heuristics.py
index 0b0ba7cff16..69330571b61 100644
--- a/torch/_inductor/runtime/triton_heuristics.py
+++ b/torch/_inductor/runtime/triton_heuristics.py
@@ -470,6 +470,12 @@ class CachingAutotuner(KernelInterface):
         if not ASTSource:
             raise RuntimeError("Installed triton version too old, please upgrade")

+        for k in compile_meta["constants"].keys():
+            assert (
+                k in compile_meta["signature"]
+                and compile_meta["signature"][k] == "constexpr"
+            )
+
         compile_args = (
             ASTSource(
                 self.fn,
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP][inductor][5/N] triton support post-#5512, fix 1 and None handling"


This fixes handling for "1" and "None" args. TL;DR: triton_meta["constants"] (which is passed to ASTSource) should be a map of {"kwarg_name": "constant value"} for values which are tl.constexpr, or have a value of 1 or None (i.e. "specialized" constants). For constant args, triton_meta["signature"][arg_name] should be "constexpr" (even for specialized constants).

In a bit more detail:
* Triton commit 5220 actually removes "1" and "None" args from the "constants" argument to ASTSource, but adds tl.constexpr args into the signature
* Triton commit 5512 brings "1" and "None" args back into the "constants" argument, and leaves tl.constexpr args in the signature. But now, the "1" and "None" args in the signature are marked as "constexpr".

So if we want to implement support for the range of triton commits labeled `TritonAttrsDescriptorVersion.V3_BACKENDS_TUPLE`, this is something we should keep in mind.

...

To test this, I added this change to triton_heuristics.py:

```diff
diff --git a/torch/_inductor/runtime/triton_heuristics.py b/torch/_inductor/runtime/triton_heuristics.py
index 0b0ba7cff16..69330571b61 100644
--- a/torch/_inductor/runtime/triton_heuristics.py
+++ b/torch/_inductor/runtime/triton_heuristics.py
@@ -470,6 +470,12 @@ class CachingAutotuner(KernelInterface):
         if not ASTSource:
             raise RuntimeError("Installed triton version too old, please upgrade")

+        for k in compile_meta["constants"].keys():
+            assert (
+                k in compile_meta["signature"]
+                and compile_meta["signature"][k] == "constexpr"
+            )
+
         compile_args = (
             ASTSource(
                 self.fn,
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP][inductor][5/N] triton support post-#5512, fix 1 and None handling"


This fixes handling for "1" and "None" args. TL;DR: triton_meta["constants"] (which is passed to ASTSource) should be a map of {"kwarg_name": "constant value"} for values which are tl.constexpr, or have a value of 1 or None (i.e. "specialized" constants). For constant args, triton_meta["signature"][arg_name] should be "constexpr" (even for specialized constants).

In a bit more detail:
* Triton commit 5220 actually removes "1" and "None" args from the "constants" argument to ASTSource, but adds tl.constexpr args into the signature
* Triton commit 5512 brings "1" and "None" args back into the "constants" argument, and leaves tl.constexpr args in the signature. But now, the "1" and "None" args in the signature are marked as "constexpr".

So if we want to implement support for the range of triton commits labeled `TritonAttrsDescriptorVersion.V3_BACKENDS_TUPLE`, this is something we should keep in mind.

...

To test this, I added this change to triton_heuristics.py:

```diff
diff --git a/torch/_inductor/runtime/triton_heuristics.py b/torch/_inductor/runtime/triton_heuristics.py
index 0b0ba7cff16..69330571b61 100644
--- a/torch/_inductor/runtime/triton_heuristics.py
+++ b/torch/_inductor/runtime/triton_heuristics.py
@@ -470,6 +470,12 @@ class CachingAutotuner(KernelInterface):
         if not ASTSource:
             raise RuntimeError("Installed triton version too old, please upgrade")

+        for k in compile_meta["constants"].keys():
+            assert (
+                k in compile_meta["signature"]
+                and compile_meta["signature"][k] == "constexpr"
+            )
+
         compile_args = (
             ASTSource(
                 self.fn,
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP][inductor][5/N] triton support post-#5512, fix 1 and None handling"


This fixes handling for "1" and "None" args. TL;DR: triton_meta["constants"] (which is passed to ASTSource) should be a map of {"kwarg_name": "constant value"} for values which are tl.constexpr, or have a value of 1 or None (i.e. "specialized" constants). For constant args, triton_meta["signature"][arg_name] should be "constexpr" (even for specialized constants).

In a bit more detail:
* Triton commit 5220 actually removes "1" and "None" args from the "constants" argument to ASTSource, but adds tl.constexpr args into the signature
* Triton commit 5512 brings "1" and "None" args back into the "constants" argument, and leaves tl.constexpr args in the signature. But now, the "1" and "None" args in the signature are marked as "constexpr".

So if we want to implement support for the range of triton commits labeled `TritonAttrsDescriptorVersion.V3_BACKENDS_TUPLE`, this is something we should keep in mind.

...

To test this, I added this change to triton_heuristics.py:

```diff
diff --git a/torch/_inductor/runtime/triton_heuristics.py b/torch/_inductor/runtime/triton_heuristics.py
index 0b0ba7cff16..69330571b61 100644
--- a/torch/_inductor/runtime/triton_heuristics.py
+++ b/torch/_inductor/runtime/triton_heuristics.py
@@ -470,6 +470,12 @@ class CachingAutotuner(KernelInterface):
         if not ASTSource:
             raise RuntimeError("Installed triton version too old, please upgrade")

+        for k in compile_meta["constants"].keys():
+            assert (
+                k in compile_meta["signature"]
+                and compile_meta["signature"][k] == "constexpr"
+            )
+
         compile_args = (
             ASTSource(
                 self.fn,
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP][inductor][5/N] triton support post-#5512, fix 1 and None handling"


This fixes handling for "1" and "None" args. TL;DR: triton_meta["constants"] (which is passed to ASTSource) should be a map of {"kwarg_name": "constant value"} for values which are tl.constexpr, or have a value of 1 or None (i.e. "specialized" constants). For constant args, triton_meta["signature"][arg_name] should be "constexpr" (even for specialized constants).

In a bit more detail:
* Triton commit 5220 actually removes "1" and "None" args from the "constants" argument to ASTSource, but adds tl.constexpr args into the signature
* Triton commit 5512 brings "1" and "None" args back into the "constants" argument, and leaves tl.constexpr args in the signature. But now, the "1" and "None" args in the signature are marked as "constexpr".

So if we want to implement support for the range of triton commits labeled `TritonAttrsDescriptorVersion.V3_BACKENDS_TUPLE`, this is something we should keep in mind.

...

To test this, I added this change to triton_heuristics.py:

```diff
diff --git a/torch/_inductor/runtime/triton_heuristics.py b/torch/_inductor/runtime/triton_heuristics.py
index 0b0ba7cff16..69330571b61 100644
--- a/torch/_inductor/runtime/triton_heuristics.py
+++ b/torch/_inductor/runtime/triton_heuristics.py
@@ -470,6 +470,12 @@ class CachingAutotuner(KernelInterface):
         if not ASTSource:
             raise RuntimeError("Installed triton version too old, please upgrade")

+        for k in compile_meta["constants"].keys():
+            assert (
+                k in compile_meta["signature"]
+                and compile_meta["signature"][k] == "constexpr"
+            )
+
         compile_args = (
             ASTSource(
                 self.fn,
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP][inductor][5/N] triton support post-#5512, fix 1 and None handling"


This fixes handling for "1" and "None" args. TL;DR: triton_meta["constants"] (which is passed to ASTSource) should be a map of {"kwarg_name": "constant value"} for values which are tl.constexpr, or have a value of 1 or None (i.e. "specialized" constants). For constant args, triton_meta["signature"][arg_name] should be "constexpr" (even for specialized constants).

In a bit more detail:
* Triton commit 5220 actually removes "1" and "None" args from the "constants" argument to ASTSource, but adds tl.constexpr args into the signature
* Triton commit 5512 brings "1" and "None" args back into the "constants" argument, and leaves tl.constexpr args in the signature. But now, the "1" and "None" args in the signature are marked as "constexpr".

So if we want to implement support for the range of triton commits labeled `TritonAttrsDescriptorVersion.V3_BACKENDS_TUPLE`, this is something we should keep in mind.

...

To test this, I added this change to triton_heuristics.py:

```diff
diff --git a/torch/_inductor/runtime/triton_heuristics.py b/torch/_inductor/runtime/triton_heuristics.py
index 0b0ba7cff16..69330571b61 100644
--- a/torch/_inductor/runtime/triton_heuristics.py
+++ b/torch/_inductor/runtime/triton_heuristics.py
@@ -470,6 +470,12 @@ class CachingAutotuner(KernelInterface):
         if not ASTSource:
             raise RuntimeError("Installed triton version too old, please upgrade")

+        for k in compile_meta["constants"].keys():
+            assert (
+                k in compile_meta["signature"]
+                and compile_meta["signature"][k] == "constexpr"
+            )
+
         compile_args = (
             ASTSource(
                 self.fn,
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP][inductor][5/N] triton support post-#5512, fix 1 and None handling"


This fixes handling for "1" and "None" args. TL;DR: triton_meta["constants"] (which is passed to ASTSource) should be a map of {"kwarg_name": "constant value"} for values which are tl.constexpr, or have a value of 1 or None (i.e. "specialized" constants). For constant args, triton_meta["signature"][arg_name] should be "constexpr" (even for specialized constants).

In a bit more detail:
* Triton commit 5220 actually removes "1" and "None" args from the "constants" argument to ASTSource, but adds tl.constexpr args into the signature
* Triton commit 5512 brings "1" and "None" args back into the "constants" argument, and leaves tl.constexpr args in the signature. But now, the "1" and "None" args in the signature are marked as "constexpr".

So if we want to implement support for the range of triton commits labeled `TritonAttrsDescriptorVersion.V3_BACKENDS_TUPLE`, this is something we should keep in mind.

...

To test this, I added this change to triton_heuristics.py:

```diff
diff --git a/torch/_inductor/runtime/triton_heuristics.py b/torch/_inductor/runtime/triton_heuristics.py
index 0b0ba7cff16..69330571b61 100644
--- a/torch/_inductor/runtime/triton_heuristics.py
+++ b/torch/_inductor/runtime/triton_heuristics.py
@@ -470,6 +470,12 @@ class CachingAutotuner(KernelInterface):
         if not ASTSource:
             raise RuntimeError("Installed triton version too old, please upgrade")

+        for k in compile_meta["constants"].keys():
+            assert (
+                k in compile_meta["signature"]
+                and compile_meta["signature"][k] == "constexpr"
+            )
+
         compile_args = (
             ASTSource(
                 self.fn,
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP][inductor][5/N] triton support post-#5512, fix 1 and None handling"


This fixes handling for "1" and "None" args. TL;DR: triton_meta["constants"] (which is passed to ASTSource) should be a map of {"kwarg_name": "constant value"} for values which are tl.constexpr, or have a value of 1 or None (i.e. "specialized" constants). For constant args, triton_meta["signature"][arg_name] should be "constexpr" (even for specialized constants).

In a bit more detail:
* Triton commit 5220 actually removes "1" and "None" args from the "constants" argument to ASTSource, but adds tl.constexpr args into the signature
* Triton commit 5512 brings "1" and "None" args back into the "constants" argument, and leaves tl.constexpr args in the signature. But now, the "1" and "None" args in the signature are marked as "constexpr".

So if we want to implement support for the range of triton commits labeled `TritonAttrsDescriptorVersion.V3_BACKENDS_TUPLE`, this is something we should keep in mind.

...

To test this, I added this change to triton_heuristics.py:

```diff
diff --git a/torch/_inductor/runtime/triton_heuristics.py b/torch/_inductor/runtime/triton_heuristics.py
index 0b0ba7cff16..69330571b61 100644
--- a/torch/_inductor/runtime/triton_heuristics.py
+++ b/torch/_inductor/runtime/triton_heuristics.py
@@ -470,6 +470,12 @@ class CachingAutotuner(KernelInterface):
         if not ASTSource:
             raise RuntimeError("Installed triton version too old, please upgrade")

+        for k in compile_meta["constants"].keys():
+            assert (
+                k in compile_meta["signature"]
+                and compile_meta["signature"][k] == "constexpr"
+            )
+
         compile_args = (
             ASTSource(
                 self.fn,
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP][inductor][5/N] triton support post-#5512, fix 1 and None handling"


This fixes handling for "1" and "None" args. TL;DR: triton_meta["constants"] (which is passed to ASTSource) should be a map of {"kwarg_name": "constant value"} for values which are tl.constexpr, or have a value of 1 or None (i.e. "specialized" constants). For constant args, triton_meta["signature"][arg_name] should be "constexpr" (even for specialized constants).

In a bit more detail:
* Triton commit 5220 actually removes "1" and "None" args from the "constants" argument to ASTSource, but adds tl.constexpr args into the signature
* Triton commit 5512 brings "1" and "None" args back into the "constants" argument, and leaves tl.constexpr args in the signature. But now, the "1" and "None" args in the signature are marked as "constexpr".

So if we want to implement support for the range of triton commits labeled `TritonAttrsDescriptorVersion.V3_BACKENDS_TUPLE`, this is something we should keep in mind.

...

To test this, I added this change to triton_heuristics.py:

```diff
diff --git a/torch/_inductor/runtime/triton_heuristics.py b/torch/_inductor/runtime/triton_heuristics.py
index 0b0ba7cff16..69330571b61 100644
--- a/torch/_inductor/runtime/triton_heuristics.py
+++ b/torch/_inductor/runtime/triton_heuristics.py
@@ -470,6 +470,12 @@ class CachingAutotuner(KernelInterface):
         if not ASTSource:
             raise RuntimeError("Installed triton version too old, please upgrade")

+        for k in compile_meta["constants"].keys():
+            assert (
+                k in compile_meta["signature"]
+                and compile_meta["signature"][k] == "constexpr"
+            )
+
         compile_args = (
             ASTSource(
                 self.fn,
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP][inductor][5/N] triton support post-#5512, fix 1 and None handling"


This fixes handling for "1" and "None" args. TL;DR: triton_meta["constants"] (which is passed to ASTSource) should be a map of {"kwarg_name": "constant value"} for values which are tl.constexpr, or have a value of 1 or None (i.e. "specialized" constants). For constant args, triton_meta["signature"][arg_name] should be "constexpr" (even for specialized constants).

In a bit more detail:
* Triton commit 5220 actually removes "1" and "None" args from the "constants" argument to ASTSource, but adds tl.constexpr args into the signature
* Triton commit 5512 brings "1" and "None" args back into the "constants" argument, and leaves tl.constexpr args in the signature. But now, the "1" and "None" args in the signature are marked as "constexpr".

So if we want to implement support for the range of triton commits labeled `TritonAttrsDescriptorVersion.V3_BACKENDS_TUPLE`, this is something we should keep in mind.

...

To test this, I added this change to triton_heuristics.py:

```diff
diff --git a/torch/_inductor/runtime/triton_heuristics.py b/torch/_inductor/runtime/triton_heuristics.py
index 0b0ba7cff16..69330571b61 100644
--- a/torch/_inductor/runtime/triton_heuristics.py
+++ b/torch/_inductor/runtime/triton_heuristics.py
@@ -470,6 +470,12 @@ class CachingAutotuner(KernelInterface):
         if not ASTSource:
             raise RuntimeError("Installed triton version too old, please upgrade")

+        for k in compile_meta["constants"].keys():
+            assert (
+                k in compile_meta["signature"]
+                and compile_meta["signature"][k] == "constexpr"
+            )
+
         compile_args = (
             ASTSource(
                 self.fn,
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP][inductor][5/N] triton support post-#5512, fix 1 and None handling"


This fixes handling for "1" and "None" args. TL;DR: triton_meta["constants"] (which is passed to ASTSource) should be a map of {"kwarg_name": "constant value"} for values which are tl.constexpr, or have a value of 1 or None (i.e. "specialized" constants). For constant args, triton_meta["signature"][arg_name] should be "constexpr" (even for specialized constants).

In a bit more detail:
* Triton commit 5220 actually removes "1" and "None" args from the "constants" argument to ASTSource, but adds tl.constexpr args into the signature
* Triton commit 5512 brings "1" and "None" args back into the "constants" argument, and leaves tl.constexpr args in the signature. But now, the "1" and "None" args in the signature are marked as "constexpr".

So if we want to implement support for the range of triton commits labeled `TritonAttrsDescriptorVersion.V3_BACKENDS_TUPLE`, this is something we should keep in mind.

...

To test this, I added this change to triton_heuristics.py:

```diff
diff --git a/torch/_inductor/runtime/triton_heuristics.py b/torch/_inductor/runtime/triton_heuristics.py
index 0b0ba7cff16..69330571b61 100644
--- a/torch/_inductor/runtime/triton_heuristics.py
+++ b/torch/_inductor/runtime/triton_heuristics.py
@@ -470,6 +470,12 @@ class CachingAutotuner(KernelInterface):
         if not ASTSource:
             raise RuntimeError("Installed triton version too old, please upgrade")

+        for k in compile_meta["constants"].keys():
+            assert (
+                k in compile_meta["signature"]
+                and compile_meta["signature"][k] == "constexpr"
+            )
+
         compile_args = (
             ASTSource(
                 self.fn,
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP][inductor][5/N] triton support post-#5512, fix 1 and None handling"


This fixes handling for "1" and "None" args. TL;DR: triton_meta["constants"] (which is passed to ASTSource) should be a map of {"kwarg_name": "constant value"} for values which are tl.constexpr, or have a value of 1 or None (i.e. "specialized" constants). For constant args, triton_meta["signature"][arg_name] should be "constexpr" (even for specialized constants).

In a bit more detail:
* Triton commit 5220 actually removes "1" and "None" args from the "constants" argument to ASTSource, but adds tl.constexpr args into the signature
* Triton commit 5512 brings "1" and "None" args back into the "constants" argument, and leaves tl.constexpr args in the signature. But now, the "1" and "None" args in the signature are marked as "constexpr".

So if we want to implement support for the range of triton commits labeled `TritonAttrsDescriptorVersion.V3_BACKENDS_TUPLE`, this is something we should keep in mind.

...

To test this, I added this change to triton_heuristics.py:

```diff
diff --git a/torch/_inductor/runtime/triton_heuristics.py b/torch/_inductor/runtime/triton_heuristics.py
index 0b0ba7cff16..69330571b61 100644
--- a/torch/_inductor/runtime/triton_heuristics.py
+++ b/torch/_inductor/runtime/triton_heuristics.py
@@ -470,6 +470,12 @@ class CachingAutotuner(KernelInterface):
         if not ASTSource:
             raise RuntimeError("Installed triton version too old, please upgrade")

+        for k in compile_meta["constants"].keys():
+            assert (
+                k in compile_meta["signature"]
+                and compile_meta["signature"][k] == "constexpr"
+            )
+
         compile_args = (
             ASTSource(
                 self.fn,
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP][inductor][5/N] triton support post-#5512, fix 1 and None handling"


This fixes handling for "1" and "None" args. TL;DR: triton_meta["constants"] (which is passed to ASTSource) should be a map of {"kwarg_name": "constant value"} for values which are tl.constexpr, or have a value of 1 or None (i.e. "specialized" constants). For constant args, triton_meta["signature"][arg_name] should be "constexpr" (even for specialized constants).

In a bit more detail:
* Triton commit 5220 actually removes "1" and "None" args from the "constants" argument to ASTSource, but adds tl.constexpr args into the signature
* Triton commit 5512 brings "1" and "None" args back into the "constants" argument, and leaves tl.constexpr args in the signature. But now, the "1" and "None" args in the signature are marked as "constexpr".

So if we want to implement support for the range of triton commits labeled `TritonAttrsDescriptorVersion.V3_BACKENDS_TUPLE`, this is something we should keep in mind.

...

To test this, I added this change to triton_heuristics.py:

```diff
diff --git a/torch/_inductor/runtime/triton_heuristics.py b/torch/_inductor/runtime/triton_heuristics.py
index 0b0ba7cff16..69330571b61 100644
--- a/torch/_inductor/runtime/triton_heuristics.py
+++ b/torch/_inductor/runtime/triton_heuristics.py
@@ -470,6 +470,12 @@ class CachingAutotuner(KernelInterface):
         if not ASTSource:
             raise RuntimeError("Installed triton version too old, please upgrade")

+        for k in compile_meta["constants"].keys():
+            assert (
+                k in compile_meta["signature"]
+                and compile_meta["signature"][k] == "constexpr"
+            )
+
         compile_args = (
             ASTSource(
                 self.fn,
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

SamGinzburg reviewed (approved):
I think this looks good to me, not sure why the CPU smoke test was failing---rerunning it now. If it was just flakey I think this is good.

davidberard98 commented:
@pytorchbot merge -r

davidberard98 committed:
Update on "[inductor][5/N] triton support post-#5512, fix 1 and None handling"


This fixes handling for "1" and "None" args with new Triton versions. TL;DR: triton_meta["constants"] (which is passed to ASTSource) should be a map of {"kwarg_name": constant_value} for values which are tl.constexpr, or have a value of 1 or None (i.e. "specialized" constants). For constant args, triton_meta["signature"][arg_name] should be "constexpr" (even for specialized constants).

Note: This adds support for Triton versions after 5512; but not for versions in between 5220 and 5512 (i.e. `TritonAttrsDescriptorVersion.V3_BACKENDS_TUPLE`). There's a completely different format for constants/signature in the commit range in between.

To test: I ran `test_torchinductor.py` and `test_triton_kernels.py` with the main branch of triton (~jan 27). The only failing tests are aoti-related tests (which need to be fixed as a follow-up), and test_mutable_custom_op_fixed_layout2_cuda (which is failing with or without the new triton version on my machine); additionally, the split-scan/split-reduction kernels rely on https://github.com/triton-lang/triton/pull/5723.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

davidberard98 committed:
lint after bad merge conflict resolution on "[inductor][5/N] triton support post-#5512, fix 1 and None handling"


This fixes handling for "1" and "None" args with new Triton versions. TL;DR: triton_meta["constants"] (which is passed to ASTSource) should be a map of {"kwarg_name": constant_value} for values which are tl.constexpr, or have a value of 1 or None (i.e. "specialized" constants). For constant args, triton_meta["signature"][arg_name] should be "constexpr" (even for specialized constants).

Note: This adds support for Triton versions after 5512; but not for versions in between 5220 and 5512 (i.e. `TritonAttrsDescriptorVersion.V3_BACKENDS_TUPLE`). There's a completely different format for constants/signature in the commit range in between.

To test: I ran `test_torchinductor.py` and `test_triton_kernels.py` with the main branch of triton (~jan 27). The only failing tests are aoti-related tests (which need to be fixed as a follow-up), and test_mutable_custom_op_fixed_layout2_cuda (which is failing with or without the new triton version on my machine); additionally, the split-scan/split-reduction kernels rely on https://github.com/triton-lang/triton/pull/5723.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

davidberard98 commented:
@pytorchbot merge

davidberard98 commented:
@pytorchbot merge
